storage/mysql/mysql.go

```diff
@@ -159,3 +182,126 @@ func (s *Storage) ReadEntryBundle(ctx context.Context, index uint64) ([]byte, er
 	var entryBundle []byte
 	return entryBundle, row.Scan(&entryBundle)
 }
+
+// Add is the entrypoint for adding entries to a sequencing log.
+func (s *Storage) Add(ctx context.Context, entry *tessera.Entry) (uint64, error) {
+	// TODO: Return index if the value is already stored.
+
+	return s.queue.Add(ctx, entry)()
+}
+
+// sequenceBatch writes the entries from the provided batch into the entry bundle files of the log.
+//
+// This func starts filling entries bundles at the next available slot in the log, ensuring that the
+// sequenced entries are contiguous from the zeroth entry (i.e left-hand dense).
+// We try to minimise the number of partially complete entry bundles by writing entries in chunks rather
+// than one-by-one.
+//
+// TODO: Separate sequencing and integration for better performance.
+func (s *Storage) sequenceBatch(ctx context.Context, entries []*tessera.Entry) error {
+	// Return when there is no entry to sequence.
+	if len(entries) == 0 {
+		return nil
+	}
+
+	// Get a Tx for making transaction requests.
+	tx, err := s.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	// Defer a rollback in case anything fails.
+	defer func() {
+		if err := tx.Rollback(); err != nil && err != sql.ErrTxDone {
+			klog.Errorf("Failed to rollback in sequenceBatch: %v", err)
+		}
+	}()
+
+	// Get tree size from checkpoint.
+	row := tx.QueryRowContext(ctx, selectCheckpointByIDSQL, checkpointID)
+	if err := row.Err(); err != nil {
+		return err
+	}
+	var rawCheckpoint []byte
+	if err := row.Scan(&rawCheckpoint); err != nil {
+		return fmt.Errorf("failed to read checkpoint: %w", err)
+	}
+	checkpoint, err := s.parseCheckpoint(rawCheckpoint)
+	if err != nil {
+		return fmt.Errorf("failed to verify checkpoint: %w", err)
+	}
```

Comment:
It _might_ be worth considering an optimization to avoid parsing and verifying the sigs each time. Two options spring to mind:
 - storing the size in a column
 - a hackier version of parsing the checkpoint where we just pull out the second line and don't verify the sigs

We should leave this for now and approach this with benchmarking later.

Reply:
Good idea. Added a TODO and linked to #21 .